### PR TITLE
fix: Add validation format check for SDK key

### PR DESCRIPTION
**Pull request by jsonbailey**
## Fix SDK Key Format Validation to Prevent Logging Invalid Keys

This PR implements SDK key format validation to prevent logging of potentially sensitive invalid SDK keys, aligning the Python SDK with the .NET SDK implementation.

**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission guidelines](../blob/main/CONTRIBUTING.md#submitting-pull-requests)
- [x] I have validated my changes against all supported platform versions

**Related issues**

Addresses security concern where invalid SDK keys could be logged, potentially exposing sensitive information. Based on similar validation implemented in [dotnet-core PR #163](https://github.com/launchdarkly/dotnet-core/pull/163) and [PR #162](https://github.com/launchdarkly/dotnet-core/pull/162).

**Describe the solution you've provided**

### Key Changes:
1. **New validation function**: Added `is_valid_sdk_key_format()` in `ldclient/impl/util.py` that validates SDK keys using:
   - Regex pattern `^[-a-zA-Z0-9._]+$` (matching .NET SDK)
   - Maximum length of 8192 characters
   - Allows empty/None keys for offline mode

2. **Silent failure behavior**: Modified `Config` constructor and `copy_with_new_sdk_key()` to silently set invalid keys to `None` instead of raising `ValueError` exceptions

3. **Comprehensive test coverage**: Added tests for valid keys, invalid keys, edge cases (empty/None/non-string), and maximum length validation

### Security Benefit:
Invalid SDK keys are now prevented from entering the system, eliminating the risk of them being logged elsewhere in the codebase.

**Critical Review Points:**
- ⚠️ **Behavioral Change**: This changes from raising exceptions to silent failure - verify this won't break existing error handling
- ⚠️ **Regex Pattern**: Confirm `^[-a-zA-Z0-9._]+$` exactly matches the .NET SDK implementation
- ⚠️ **None Handling**: Ensure all SDK code paths properly handle `None` SDK keys

**Describe alternatives you've considered**

- **Continue raising exceptions**: Rejected because it doesn't prevent logging and breaks the goal of silent validation
- **Log validation errors**: Rejected because it would still potentially log the invalid key value
- **Use HTTP header validation regex**: Rejected to maintain consistency with .NET SDK pattern

**Additional context**

- Link to Devin run: https://app.devin.ai/sessions/c9e47d6d2dcb4ad3a56648b1c35cc563
- Requested by @jsonbailey
- Includes branch naming with sdk-1460 as requested
- Local testing shows 475 tests passing with only unrelated infrastructure failures (Consul/DynamoDB connection issues)

---

> [!NOTE]
> Add strict SDK key validation (length/characters) and integrate into Config; standardize constants; add comprehensive tests.
> 
> - **SDK key validation**:
>   - Add `validate_sdk_key_format` in `ldclient/impl/util.py` enforcing allowed characters and max length (8192), returning `''` on invalid and logging warnings.
>   - Use validation in `Config` (`ldclient/config.py`) to set `__sdk_key`; update `copy_with_new_sdk_key` behavior accordingly.
> - **Logging/validation tweaks**:
>   - Adjust `_validate` in `Config` to warn on missing/blank SDK key with updated message.
> - **Refactors**:
>   - Standardize constants in `impl/util.py` (e.g., `_retryable_statuses` → `_RETRYABLE_STATUSES`) and update references.
> - **Tests**:
>   - Add `ldclient/testing/impl/test_util.py` covering valid/invalid/non-string/empty/None/max-length SDK keys.
>   - Expand `ldclient/testing/test_config.py` to verify SDK key validation, copy behavior, and max length handling.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6ec6770153c3712aac9a0f4587647af5208177da. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

BEGIN_COMMIT_OVERRIDE
fix: Add validation format check for SDK key
END_COMMIT_OVERRIDE

### CODE REVIEW

File: ldclient/impl/util.py [hunk 1]
```diff
@@ -53,6 +53,25 @@ def validate_application_value(value: Any, name: str, logger: logging.Logger) ->
     return value
 
 
+def validate_sdk_key(sdk_key: str, logger: logging.Logger) -> bool:
+    """
+    Validate that an SDK key contains only characters that are valid for HTTP headers.
+    Returns True if valid, False if invalid. Logs a generic error message for invalid keys.
+    """
```
Comment (keelerm84) [inline ldclient/impl/util.py:60]: Missing :param <param>: and :return: style comments.

File: ldclient/impl/util.py [hunk 2]
```diff
@@ -53,6 +53,25 @@ def validate_application_value(value: Any, name: str, logger: logging.Logger) ->
...
+    if not isinstance(sdk_key, str):
+        logger.warning("SDK key must be a string")
+        return False
+    
+    if sdk_key == '':
+        return True  # Empty keys are handled separately in _validate()
+    
+    if re.search(r"[^\x21-\x7E]", sdk_key):
```
Comment (keelerm84) [inline ldclient/impl/util.py:68]: That isn't what we went with in C# is it?

File: ldclient/config.py [hunk 3]
```diff
@@ -261,6 +261,9 @@ def __init__(
         :param omit_anonymous_contexts: Sets whether anonymous contexts should be omitted from index and identify events.
         :param payload_filter_key: The payload filter is used to selectively limited the flags and segments delivered in the data source payload.
         """
+        if sdk_key and not validate_sdk_key(sdk_key, log):
+            raise ValueError("SDK key contains invalid characters")
```
Comment (keelerm84) [inline ldclient/config.py:265]: Aside from the excellent point above, do we know for sure that an invalid SDK key caused a raised exception in python? Can't remember if you told me you confirmed that or not, so apologies if I make you repeat yourself.

**Review by keelerm84 — CHANGES_REQUESTED** [review]

File: ldclient/config.py [hunk 4]
```diff
@@ -302,6 +302,7 @@ def __init__(
 
     def copy_with_new_sdk_key(self, new_sdk_key: str) -> 'Config':
         """Returns a new ``Config`` instance that is the same as this one, except for having a different SDK key.
+        The key will not be updated if the provided key contains invalid characters.
```
[the remark this answers is not in the document]
Comment (jsonbailey) [reply ldclient/config.py:309]: This is addressed in a separate PR to keep this one focused.

File: ldclient/impl/util.py [hunk 5]
```diff
@@ -46,13 +51,38 @@ def validate_application_value(value: Any, name: str, logger: logging.Logger) ->
...
+    
+    if len(sdk_key) > _MAX_SDK_KEY_LENGTH:
+        logger.warning('SDK key was longer than %d characters and was discarded' % _MAX_SDK_KEY_LENGTH)
+        return ""
+    
+    if _VALID_CHARACTERS_REGEX.search(sdk_key):
+        logger.warning('SDK key contained invalid characters and was discarded')
+        return ""
+    
+    return sdk_key
```
[the remark this answers is not in the document]
Comment (jsonbailey) [reply ldclient/impl/util.py:80]: I updated the config validation logic to remove the None check.

[the remark this answers is not in the document]
Comment (jsonbailey) [reply ldclient/impl/util.py:80]: Both None and "" are valid in offline mode. The validation method on the config didn't prevent anything, it only logged a warning if a blank or missing sdk key was provided.

**Review by keelerm84 — APPROVED** [review]
